File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteMonthlyDayScheduleRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.firebase.json.MonthlyDayScheduleJson;
import com.krystianwsul.checkme.firebase.json.ScheduleWrapper;

import junit.framework.Assert;

public class RemoteMonthlyDayScheduleRecord extends RemoteScheduleRecord {
    RemoteMonthlyDayScheduleRecord(@NonNull String id, @NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(id, remoteTaskRecord, scheduleWrapper);
    }

    RemoteMonthlyDayScheduleRecord(@NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(remoteTaskRecord, scheduleWrapper);
    }

    @NonNull
    private MonthlyDayScheduleJson getMonthlyDayScheduleJson() {
        MonthlyDayScheduleJson monthlyDayScheduleJson = mScheduleWrapper.monthlyDayScheduleJson;
        Assert.assertTrue(monthlyDayScheduleJson != null);

        return monthlyDayScheduleJson;
    }

    @Override
    public long getStartTime() {
        return getMonthlyDayScheduleJson().getStartTime();
    }

    @Nullable
    @Override
    public Long getEndTime() {
        return getMonthlyDayScheduleJson().getEndTime();
    }

    public int getDayOfMonth() {
        return getMonthlyDayScheduleJson().getDayOfMonth();
    }

    public boolean getBeginningOfMonth() {
        return getMonthlyDayScheduleJson().getBeginningOfMonth();
    }

    @Nullable
    public String getCustomTimeId() {
        return getMonthlyDayScheduleJson().getCustomTimeId();
    }

    @Nullable
    public Integer getHour() {
        return getMonthlyDayScheduleJson().getHour();
    }

    @Nullable
    public Integer getMinute() {
        return getMonthlyDayScheduleJson().getMinute();
    }

    public void setEndTime(long endTime) {
        Assert.assertTrue(getEndTime() == null);

        getMonthlyDayScheduleJson().setEndTime(endTime);
        addValue(getKey() + "/monthlyDayScheduleJson/endTime", endTime);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteMonthlyDayScheduleRecord.kt
package com.krystianwsul.checkme.firebase.records


import com.krystianwsul.common.firebase.json.ScheduleWrapper
import com.krystianwsul.common.utils.RemoteCustomTimeId


class RemoteMonthlyDayScheduleRecord<T : RemoteCustomTimeId> : RemoteScheduleRecord<T> {

    private val monthlyDayScheduleJson by lazy { createObject.monthlyDayScheduleJson!! }

    val dayOfMonth by lazy { monthlyDayScheduleJson.dayOfMonth }

    val beginningOfMonth by lazy { monthlyDayScheduleJson.beginningOfMonth }

    override val customTimeId by lazy { monthlyDayScheduleJson.customTimeId?.let { remoteTaskRecord.getRemoteCustomTimeId(it) } }

    val hour by lazy { monthlyDayScheduleJson.hour }

    val minute by lazy { monthlyDayScheduleJson.minute }

    constructor(id: String, remoteTaskRecord: RemoteTaskRecord<T>, scheduleWrapper: ScheduleWrapper) : super(id, remoteTaskRecord, scheduleWrapper)

    constructor(remoteTaskRecord: RemoteTaskRecord<T>, scheduleWrapper: ScheduleWrapper) : super(remoteTaskRecord, scheduleWrapper)

    override val startTime by lazy { monthlyDayScheduleJson.startTime }

    override var endTime
        get() = monthlyDayScheduleJson.endTime
        set(value) {
            if (value == monthlyDayScheduleJson.endTime)
                return

            monthlyDayScheduleJson.endTime = value
            addValue("$key/monthlyDayScheduleJson/endTime", value)
        }

    override fun deleteFromParent() = check(remoteTaskRecord.remoteMonthlyDayScheduleRecords.remove(id) == this)
}
